src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,545 @@
+{
+	if (LocalTableJoinPolicy == LOCAL_JOIN_POLICY_NEVER)
+	{
+		/* user doesn't want Citus to enable local table joins */
+		return false;
+	}
+
+	if (!ContainsLocalTableDistributedTableJoin(rangeTableList))
+	{
+		return false;
+	}
+	return true;
+}
+
+
+/*
+ * HasConstantFilterOnUniqueColumn returns true if the given rangeTableEntry has a constant
+ * filter on a unique column.
+ */
+static bool
+HasConstantFilterOnUniqueColumn(RangeTblEntry *rangeTableEntry,
+								RelationRestriction *relationRestriction)
+{
+	if (rangeTableEntry == NULL)
+	{
+		return false;
+	}
+	List *baseRestrictionList = relationRestriction->relOptInfo->baserestrictinfo;
+	List *restrictClauseList = get_all_actual_clauses(baseRestrictionList);
+	if (ContainsFalseClause(restrictClauseList))
+	{
+		/* If there is a WHERE FALSE, we consider it as a constant filter. */
+		return true;
+	}
+	List *rteEqualityColumnsNos =
+		FetchEqualityAttrNumsForRTE((Node *) restrictClauseList);
+
+	List *uniqueIndexColumnsList = ExecuteFunctionOnEachTableIndex(rangeTableEntry->relid,
+																   AppendUniqueIndexColumnsToList);
+	IndexColumns *indexColumns = NULL;
+	foreach_ptr(indexColumns, uniqueIndexColumnsList)
+	{
+		List *uniqueIndexColumnNos = indexColumns->indexColumnNos;
+		if (FirstIsSuperSetOfSecond(rteEqualityColumnsNos,
+									uniqueIndexColumnNos))
+		{
+			return true;
+		}
+	}
+	return false;
+}
+
+
+/*
+ * FirstIsSuperSetOfSecond returns true if the first int List
+ * contains every element of the second int List.
+ */
+static bool
+FirstIsSuperSetOfSecond(List *firstIntList, List *secondIntList)
+{
+	int curInt = 0;
+	foreach_int(curInt, secondIntList)
+	{
+		if (!list_member_int(firstIntList, curInt))
+		{
+			return false;
+		}
+	}
+	return true;
+}
+
+
+/*
+ * AppendUniqueIndexColumnsToList adds the given index's column numbers if it is a
+ * unique index.
+ */
+static void
+AppendUniqueIndexColumnsToList(Form_pg_index indexForm, List **uniqueIndexGroups)
+{
+	if (indexForm->indisunique || indexForm->indisprimary)
+	{
+		IndexColumns *indexColumns = palloc0(sizeof(IndexColumns));
+		List *uniqueIndexes = NIL;
+		for (int i = 0; i < indexForm->indkey.dim1; i++)
+		{
+			uniqueIndexes = list_append_unique_int(uniqueIndexes,
+												   indexForm->indkey.values[i]);
+		}
+		if (list_length(uniqueIndexes) == 0)
+		{
+			return;
+		}
+		indexColumns->indexColumnNos = uniqueIndexes;
+		*uniqueIndexGroups = lappend(*uniqueIndexGroups, indexColumns);
+	}
+}
+
+
+/*
+ * RequiredAttrNumbersForRelation returns the required attribute numbers for
+ * the input RTE relation in order for the planning to succeed.
+ *
+ * The function could be optimized by not adding the columns that only appear
+ * WHERE clause as a filter (e.g., not a join clause).
+ */
+static List *
+RequiredAttrNumbersForRelation(RangeTblEntry *rangeTableEntry,
+							   PlannerRestrictionContext *plannerRestrictionContext)
+{
+	RelationRestriction *relationRestriction =
```

Comment:
RelationRestriction contains a RelOptInfo, which has an attr_needed array. Do you think we could implement this function using that array?

Reply:
It seems that this doesn't have constant expresisons such as local.a = 5. It does have expressions where both sides are rels such as local.a = dist.a

Reply:
We should be able to use this if we also remove the quals that are using in the subquery from the original query, but this requires some work hence we will probably do this with a separate PR.